doc/sphinx/language/core/modules.rst

```diff
@@ -422,6 +419,12 @@ The definition of :g:`N` using the module type expression :g:`SIG` with
 
    Module N : SIG' := M.
 
+.. exn:: No field named @ident in @qualid.
+
+   Raised when a :g:`with Definition @ident := @term` or :g:`with
+   Module @ident := @qualid'` clause is applied to a module type
+   @qualid which has no field named @ident.
```

Comment:
For the `with` syntax, the LHS is `@qualid`.
```suggestion
   Raised when a :n:`with Definition @ident := @term` or :n:`with
   Module @ident := @qualid` clause is applied to a module type
   :n:`@qualid` which has no field named :n:`@ident`.
```

You should use `:n:` to get hyperlinks.  You could just use the equivalent ```:n:`@with_declaration` ``` instead of repeating the syntax. This doesn't look consistent:

![image](https://user-images.githubusercontent.com/1253341/146827260-496ae019-c606-4a05-8068-96b0c19d2fb7.png)


![image](https://user-images.githubusercontent.com/1253341/146826797-02d78781-b90e-47f9-a591-1ac7f8f5f308.png)

Reply:
Thank for the `:n:`/`:g:` hints.

The first problem I have in explaining the error message is that it is difficult to explain in its more general form (i.e we can have `M with Definition N.t := ...` and the error message would say that `t` is not a field of `N` in `M`). This is why I restricted it to the simpler form when `@qualid` is `@ident` in the `with_declaration`. Should I explain the more general form?

Also, whether I explain the general form or not, another problem is that I need to refer to given instances of the different `qualid` or `ident` involved in the `@qualid @with_declaration` syntax (rather than to the generic `@qualid` or `@ident` syntax itself).

Would something like that be ok: ``Raised when a :n:`@with_declaration` clause with left-hand side :n:`ident` is applied to a module type :n:`qualid` which has no field named :n:`ident` ``.

Reply:
Maybe more understandable more like this: ``Raised when the final :n:`@ident` in the left-hand :n:`@qualid` of a  :n:`@with_declaration` is applied to a module type :n:`@qualid` that has no field named :n:`@ident` ``?

I take it the second `@qualid` is the one on the RHS of `Module` but I don't understand how the `Definition` form is used.  When I look in the source tree, it always looks like a fairly simple assignment, eg `with Definition A := X.A := X`, so probably a restricted form of `term`.  How does it apply to that case and how can it be described?

Too bad the doc says nothing about `with Definition`.

Reply:
It is necessarily the final ident in the lhs. But indeed, `` Raised when a component :n:`@ident` of the left-hand :n:`@qualid` of a  :n:`@with_declaration` is applied to a module type :n:`@qualid` that has no field named :n:`@ident` `` is not bad (one just have to understand that the `qualid` in the error message is the second of the explanation).

> When I look in the source tree, it always looks like a fairly simple assignment, eg `with Definition A := X.A := X`, so probably a restricted form of `term`.  How does it apply to that case and how can it be described?

Yes, it can be any term in general. But it does not matter for the error message. In `A.B.C with Definition C.D.E := term` only the qualid `A.B.C` and one ident among `C`, `D`, and `E` matter.

> Too bad the doc says nothing about `with Definition`.

Actually, it says something at the very beginning `A structure expression (...) can be: (...) a refined structure 𝑆 𝗐𝗂𝗍𝗁 𝑝:=𝑝′ or 𝑆 𝗐𝗂𝗍𝗁 𝑝:=𝑡:𝑇 where 𝑆 is a structure expression, 𝑝 and 𝑝′ are access paths, 𝑡 is a term and 𝑇 is the type of 𝑡.`, but I agree that it is not clear at all that this refers to the `with_declaration` syntax. This header should be rewritten using the concrete syntax rather than the formal "paper" syntax.

(Another thing which is not documented but a bit unpolished are the inlining levels.)

Reply:
Maybe you can refer to the "right-hand side" which could cover both cases.  But my understanding is a bit limited.

Reply:
> Maybe you can refer to the "right-hand side" which could cover both cases. But my understanding is a bit limited.

I would keep "left-hand side", meaning the left-hand side of `:=` in both variants of `with_declaration`. Pushed.